jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/JavaxWebSocketFrameHandler.java

```diff
@@ -278,24 +282,41 @@ public void onFrame(Frame frame, Callback callback)
             dataType = OpCode.UNDEFINED;
     }
 
+    public void onClose(Frame frame, Callback callback)
+    {
+        notifyOnClose(CloseStatus.getCloseStatus(frame), callback);
+    }
+
     @Override
     public void onClosed(CloseStatus closeStatus, Callback callback)
     {
+        notifyOnClose(closeStatus, callback);
+        container.notifySessionListeners((listener) -> listener.onJavaxWebSocketSessionClosed(session));
+    }
+
+    private void notifyOnClose(CloseStatus closeStatus, Callback callback)
+    {
+        // Make sure onClose is only notified once.
+        if (!closeNotified.compareAndSet(false, true))
```

Comment:
How can this happen?  Is this a problem with Core?

Reply:
No the websocket-core `onClosed` event means the websocket is fully closed. @sbordet  wanted the ability to send alternate close responses from the API `onClose` events, so now the `onClose` event at the API level can be triggered in two ways: 

Receiving a close frame (in which case the websocket is still out open to send a custom close response).

Or by the core `onClosed` event (if we never received the close frame but still get closed because of an error or something).

So we call `notifyOnClose` from both these places but we only want to actually call the API `onClose` event once.